src/NHibernate/Util/AsyncLock.cs

```diff
@@ -0,0 +1,47 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace NHibernate.Util
+{
+	public class AsyncLock
+	{
+		private readonly AsyncSemaphore semaphore;
+		private readonly Task<Releaser> releaser;
+
+		public AsyncLock()
+		{
+			semaphore = new AsyncSemaphore(1);
+			releaser = Task.FromResult(new Releaser(this));
+		}
+
+		public Task<Releaser> LockAsync()
```

Comment:
The lack of cancellation token here is intentional I think, and should not be "fixed".